Report correct type information for inlined types

hubertp opened the pull request:
### Pull Request Description

This change reconstructs full type signature from inlined types, if possible.
Also fixed a bug where a pre-mature decomposition of type arguments meant that inline type of arguments was discarded for function types.

Closes  #12519.

After this change we correctly show the return type for `Table.filter`:
![Screenshot from 2025-03-28 17-42-23](https://github.com/user-attachments/assets/6f824193-69a3-423d-aa37-96c21392ed94)


### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

hubertp committed:
Report function types for arguments in suggestions

Pre-mature decomposition of type arguments meant that inline type of
arguments was discarded.
Return type in the inline type is still wrong, as demonstrated in the
added test-case.

hubertp committed:
Reconstruct full type signature from inlined types

If type are inlined in arguments, it is possible to reconstruct full
type signature.
That also allows us keep the existing logic in passes that rely on that
information, like suggestions builder.

hubertp committed:
Support methods with partially inlined types

The code reverts to the old logic of inferring individual argument's
types in suggestions when not all types are provided.

hubertp committed:
One more corner case

hubertp committed:
Support types with declared exceptions

Another corner case.

hubertp committed:
Merge branch 'develop' into wip/hubert/12519-inline-types

4e6 commented on engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/resolve/TypeSignatures.scala:
```diff
@@ -143,9 +143,23 @@ case object TypeSignatures extends IRPass {
+                  new MetadataPair(this, Signature(typeFun))
+                )
+              }
+            Some(newMethod)
```
Please update `TypeSignaturesTest` with inline signatures test

hubertp replied:
I will create follow up work where we attempt to be more complete - currently we require `Standard.Base.Any.Any` to be resolvable (i.e. imported) to avoid no symbol found errors. In the case of IDE that is fine, as one typically import `Standard.Base`, but we could remove that limitation.

hubertp committed:
Support conversion methods with partial types

Fixes failing tests

hubertp committed:
Merge branch 'develop' into wip/hubert/12519-inline-types

hubertp commented:
I'm partially affected by a recent change https://github.com/enso-org/enso/pull/12263. This PR needs more work unfortunately.

Akirathan commented:
> I'm partially affected by a recent change https://github.com/enso-org/enso/pull/12263. This PR needs more work unfortunately.

PR #12263 was merged 2 weeks ago. This PR was created 5 days ago, with the oldest commit being 5 days ago. So you started experimenting with this functionality more than 2 weeks ago? Anyway, feel free to DM me.

hubertp committed:
Fix infinite loop in Bindings Map

Also, avoid spurious errors related to missing `Standard.Base.Any.Any`.
The latter means that inferring types from partially inlined types is
only possible if Any is resolvable.

hubertp commented:
> > I'm partially affected by a recent change #12263. This PR needs more work unfortunately.
> 
> PR #12263 was merged 2 weeks ago. This PR was created 5 days ago, with the oldest commit being 5 days ago. So you started experimenting with this functionality more than 2 weeks ago? Anyway, feel free to DM me.

I might have had an old develop from which I branched off.
